2021_scipy_tutorial/fairness-in-AI-systems-instructors.ipynb

```diff
@@ -0,0 +1,4620 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 203
+        },
+        "id": "CEUVENjeZqJF",
+        "outputId": "c4dfac39-174f-429b-d762-7bdab15d0098"
+      },
+      "outputs": [],
+      "source": [
+        "# Save the predictions and report the disagregated metrics\n",
+        "# of the exponantiated gradient model\n",
+        "Y_expgrad_exercise = expgrad_exercise.predict(X_test)\n",
+        "mf_expgrad_exercise = MetricFrame(\n",
+        "    metrics=metrics_with_err_bars,\n",
+        "    y_true=Y_test,\n",
+        "    y_pred=Y_expgrad_exercise,\n",
+        "    sensitive_features=A_test\n",
+        ")\n",
+        "mf_expgrad_exercise.by_group[metrics_to_display]"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 203
+        },
+        "id": "Q1nPfHpWRwWZ",
+        "outputId": "927aa5a0-5679-41e1-f339-45712883e935"
+      },
+      "outputs": [],
+      "source": [
+        "# Compare with the disaggregated metric values of the\n",
+        "# exponentiated gradient model based on logistic regression\n",
+        "metricframe_reductions.by_group[metrics_to_display]"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "_Qkfth4ZZBQV"
+      },
+      "outputs": [],
+      "source": [
+        "#Y_expgrad_exercise = expgrad_exercise.predict(X_test)\n",
+        "#mf_expgrad_exercise = MetricFrame(\n",
+        "#    metrics=________,\n",
+        "#    y_true=Y_test,\n",
+        "#    y_pred=_______,\n",
+        "#    sensitive_features=________\n",
+        "#)\n",
+        "#mf_expgrad_exercise.______[metrics_to_display]"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Boo771yFUyJ7"
+      },
+      "source": [
+        "3.) Next, calculate the balanced error rate and false negative rate difference of each of the inner models learned by this new `ExponentiatedGradient` classifier."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "jws7w2z6RWUM"
+      },
+      "outputs": [],
+      "source": [
+        "# Save the inner predictors of the new model\n",
+        "predictors_exercise = expgrad_exercise.predictors_"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "EAtkvVOEZqMb"
+      },
+      "outputs": [],
+      "source": [
+        "# Compute the balanced error rate and false negative rate difference for each of the predictors on the test data.\n",
+        "weighted_error_exercise = [(1 - balanced_accuracy_score(Y_test, pred.predict(X_test))) for pred in predictors_exercise]\n",
```

Comment:
`weighted_error_exercise` -> `balanced_error_exercise`
[there are 4 instances here / in collab]